util.go

```diff
@@ -127,9 +127,49 @@ func moveDeadNodes(nodes []*nodeState, gossipToTheDeadTime time.Duration) int {
 // kRandomNodes is used to select up to k random Nodes, excluding any nodes where
 // the exclude function returns true. It is possible that less than k nodes are
 // returned.
-func kRandomNodes(k int, nodes []*nodeState, exclude func(*nodeState) bool) []Node {
-	n := len(nodes)
+func kRandomNodes(k int, nodes []*nodeState, delegate NodeSelectionDelegate, exclude func(*nodeState) bool) []Node {
 	kNodes := make([]Node, 0, k)
+
+	// Filter the nodes using the delegate. This allows downstream projects
+	// to implement custom routing logics (e.g. zone-aware gossiping).
+	if delegate != nil {
+		filteredNodes := make([]*nodeState, 0, len(nodes))
+		var preferredNodes []*nodeState
```

Comment:
Given we expect preferred nodes, I think we can reduce some allocations here by pre-allocating a single array for both, something like:

```suggestion
        alloc := make([]*nodeState, 2*len(nodes))
        filteredNodes := alloc[0:0:len(nodes)]
        preferredNodes := alloc[len(nodes):len(nodes):len(nodes)]
```

Reply:
I think there's a small typo. Should be `preferredNodes := alloc[len(nodes) : len(nodes) : 2*len(nodes)]`, right?

Reply:
Yes, you're right, I've referred to some docs that were unclear on the syntax of this.

Reply:
Done in https://github.com/grafana/memberlist/pull/7/commits/ae7992860f8f49ae324728dd4990f7a982195c9a